### PR TITLE
whisper: add primitives (WIP)

### PR DESCRIPTION
**Pull request by arnetheduck**
Work-in-progress dump of initial whisper support - not for merging

- [X] payload encoding / decoding (missing 256-bit GCM mode)
- [X] missing tests
- [ ] more unit tests
- [ ] rlpx integration / glue
- [ ] peer management - send messages to joining peers, send message only once to each peer (hash-based?)
- [ ] topic-based encryption
- [ ] integration tests
- [ ] verify it works against geth, all combinations of signature and encryption

### CODE REVIEW

File: eth_p2p/rlpx_protocols/shh.nim [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+
+    capacity: int ## Max messages to keep. \
+    ## XXX: really big messages can cause excessive mem usage when using msg \
+    ##      count
+
+# Utilities --------------------------------------------------------------------
+
+proc toBE(v: uint64): array[8, byte] =
+  # return uint64 as bigendian array - for easy consumption with hash function
+  var v = cast[array[8, byte]](v)
```
Comment (yglukhov) [inline eth_p2p/rlpx_protocols/shh.nim]: No need for intermediate, you can use `unsafeAddr`

File: eth_p2p/rlpx_protocols/shh.nim [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+    proc messages(p: Peer, values: openarray[KeyValuePair]) =
+      discard
+
+    proc powRequirement(p: Peer, values: openarray[KeyValuePair]) =
+      discard
+
+    proc topicFilter(p: Peer, values: openarray[KeyValuePair]) =
+      discard
+
+if isMainModule:
```
Comment (yglukhov) [inline eth_p2p/rlpx_protocols/shh.nim]: This test will not run in CI. Why not move it to `tests/` folder?

**Comment by cheatfate** [comment]
@arnetheduck nimcrypto support both AES-GCM-128 and AES-GCM-256.
https://github.com/cheatfate/nimcrypto/issues/11

File: eth_p2p/rlpx_protocols/shh.nim (hunk 2 above)
Comment (arnetheduck) [reply eth_p2p/rlpx_protocols/shh.nim]: yeah, this is just wip.. more convenient during development (faster build), will move for sure when ready

**Comment by arnetheduck** [comment]
> Nice! Though the file is somewhat inconsistent with the rest of rlpx_protocols there because those define rlpxProtocol. Are you going to add it later?

rlpxProtocol is merely a glue layer to the underlying functionality, in my head - thus the last thing to write.. the implemented parts of whisper so far have nothing to do with it and could be used in any library or app that wants to do whisper (ie there's a libp2p whisper in the works and PSS reuses many of the data structures found in whisper) - most importantly, all of it can be unit tested separately from the complexities and whims of all that macro magic. 

keeping the responsibilities orthogonal should be an aim for any protocol we do - the less code we have hidden behind that monster, the easier it will be to isolate errors and debug issues.

**Comment by zah** [comment]
This was rebased and merged

**Pull request closed without merging**